WalletWasabi/Userfacing/AddressStringParser.cs

```diff
@@ -1,99 +1,99 @@
 using NBitcoin;
-using NBitcoin.Payment;
 using System.Diagnostics.CodeAnalysis;
-using System.Globalization;
+using WalletWasabi.Userfacing.Bip21;
 
 namespace WalletWasabi.Userfacing;
 
 public static class AddressStringParser
 {
-	public static bool TryParseBitcoinAddress(string text, Network expectedNetwork, [NotNullWhen(true)] out BitcoinUrlBuilder? url)
+	public static bool TryParse(string text, Network expectedNetwork, [NotNullWhen(true)] out Bip21UriParser.Result? result)
+		=> TryParse(text, expectedNetwork, out result, out _);
+
+	/// <summary>
+	/// Parses either a Bitcoin address or a BIP21 URI string.
+	/// </summary>
+	/// <seealso href="https://github.com/lightning/bolts/blob/master/11-payment-encoding.md"/>
+	public static bool TryParse(
+		string text,
+		Network expectedNetwork,
+		[NotNullWhen(true)] out Bip21UriParser.Result? result,
+		out string? errorMessage)
 	{
-		url = null;
+		result = null;
+		errorMessage = null;
 
 		if (text is null || expectedNetwork is null)
 		{
+			errorMessage = "Internal error.";
 			return false;
 		}
 
 		text = text.Trim();
 
-		if (text.Length is > 100 or < 20)
+		if (text == "")
 		{
+			errorMessage = "Input length is invalid.";
 			return false;
 		}
 
-		try
-		{
-			var bitcoinAddress = BitcoinAddress.Create(text, expectedNetwork);
-			url = new BitcoinUrlBuilder($"bitcoin:{bitcoinAddress}", expectedNetwork);
-			return true;
-		}
-		catch (FormatException)
+		// Too long URIs/Bitcoin address are unsupported.
+		if (text.Length > 1000)
 		{
+			errorMessage = "Input is too long.";
 			return false;
 		}
-	}
 
-	public static bool TryParseBitcoinUrl(string text, Network expectedNetwork, [NotNullWhen(true)] out BitcoinUrlBuilder? url)
-	{
-		url = null;
+		// Lightning addresses are unsupported.
+		bool isLightningAddress = text.StartsWith("lnbc", StringComparison.Ordinal) // Lightning on main.
+			|| text.StartsWith("lntb", StringComparison.Ordinal) // Lightning on testnet.
+			|| text.StartsWith("lntbs", StringComparison.Ordinal) // Lightning on signet.
+			|| text.StartsWith("lnbcrt", StringComparison.Ordinal) // Lightning on regtest
+			|| text.StartsWith("lnurl", StringComparison.Ordinal); // Lightning invoice.
 
-		if (text is null || expectedNetwork is null)
+		if (isLightningAddress)
 		{
+			errorMessage = "Lightning addresses are not supported.";
 			return false;
 		}
 
-		text = text.Trim();
-
-		if (text.Length is > 1000 or < 20)
-		{
-			return false;
-		}
+		Bip21UriParser.Error? error;
 
-		try
+		// Parse a Bitcoin address (not BIP21 URI string)
+		if (!text.StartsWith($"{Bip21UriParser.UriScheme}:", StringComparison.OrdinalIgnoreCase))
 		{
-			if (!text.StartsWith("bitcoin:", true, CultureInfo.InvariantCulture))
+			if (!Bip21UriParser.TryParseBitcoinAddressForNetwork(text, expectedNetwork, out BitcoinAddress? address))
```

Comment:
Remove the negation and switch the logic. In that case, you can remove else as well because of return.